core/api_token_api.php

```diff
@@ -0,0 +1,162 @@
+ * @return string The one way hash for the token
+ * @access public
+ */
+function api_token_hash( $p_token ) {
+	# We ignore spaces in tokens.  We inject spaces for users for readability only.
+	return sha1( str_replace( ' ', '', $p_token ) );
+}
+
+/**
+ * Validate a plain token for the specified user.
+ * @param string $p_username The user name.
+ * @param string $p_token The plain token.
+ * @return boolean true valid username and token, false otherwise.
+ * @access public
+ */
+function api_token_validate( $p_username, $p_token ) {
+	# If the supplied token doesn't look like a valid one, then fail the check w/o doing db lookups.
+	# This is likely called from code that supports both tokens and passwords.
+	if( is_blank( $p_token ) || strlen( $p_token ) != API_TOKEN_LENGTH ) {
+		return false;
+	}
+
+	$t_user_id = user_get_id_by_name( $p_username );
+
+	# If user is not found in the database, they don't have api tokens, we won't bother with worrying about
+	# auto-creation scenario here.
+	if( $t_user_id === false ) {
+		return false;
+	}
+
+	$t_encrypted_token = api_token_hash( $p_token );
+
+	$t_query = 'SELECT * FROM {api_token} WHERE user_id=' . db_param() . ' AND hash=' . db_param();
+	$t_result = db_query( $t_query, array( $t_user_id, $t_encrypted_token ) );
+
+	$t_row = db_fetch_array( $t_result );
+	if( $t_row ) {
+		api_token_touch( $t_row['id'] );
+		return true;
+	}
+
+	return false;
+}
+
+/**
+ * Get all API tokens associated with the specified user.
+ * @param integer $p_user_id The user id.
+ * @return array Array of API token rows for owned by the user, can be empty.
+ * @access public
+ */
+function api_token_get_all( $p_user_id ) {
+	$t_query = 'SELECT * FROM {api_token} WHERE user_id=' . db_param() . ' ORDER BY date_used DESC, date_created ASC';
+	$t_result = db_query( $t_query, array( $p_user_id ) );
+
+	$t_rows = array();
+	while ( ( $t_row = db_fetch_array( $t_result ) ) !== false )
+	{
+		$t_rows[] = $t_row;
+	}
+
+	return $t_rows;
+}
+
+/**
+ * Updates the last used timestamp for the api token.
+ *
+ * @param integer $p_api_token_id The token id.
+ * @return void
+ * @access public
+ */
+function api_token_touch( $p_api_token_id ) {
+	$t_date_used = db_now();
+
+	$t_query = 'UPDATE {api_token} SET date_used=' . db_param() . ' WHERE id=' . db_param();
+
+	db_query( $t_query, array( $t_date_used, $p_api_token_id ) );
+}
+
+/**
+ * Revokes the api token with the specified id
+ * @param integer $p_api_token_id The API token id.
+ * @param integer $p_user_id The user id.
+ * @return void
+ * @access public
+ */
+function api_token_revoke( $p_api_token_id, $p_user_id ) {
+	$t_query = 'DELETE FROM {api_token} WHERE id=' . db_param() . ' AND user_id = ' . db_param();
+	db_query( $t_query, array( $p_api_token_id, $p_user_id ) );
+}
+
+/**
+ * Gets the max length for the API token name.
+ * @return integer Maximum length for API token name.
+ * @access public
+ */
+function api_token_name_max_length() {
+	return 128;
```

Comment:
We implemented another approach for similar fields (e.h. user name).
Used a constant for it, define( 'DB_FIELD_SIZE_USERNAME', 255 );